chef/automate
# Add APIs to show the progress of policyfile code rollouts

### :nut_and_bolt: Description: What code changed, and why?

Add APIs to give statistics about the number of nodes having rolled out recent revisions of policies.

This work is done ahead of having a UI mockup, so it's a little bit speculative and probably will need a bit of updating to match the needs of the UI. That said, here's what it does:

* Queries the (pg) database for the 5 most recent rollouts per each node segment. Node segment is the policy name, policy group, and policy domain (Chef Server, for now) URL that groups nodes according to the code assigned to them.
* Queries the node-state database (in elastic) to aggregate node-state statistics, grouped by node segment plus policy revision. This gives us the number of nodes in each segment, the number of successes and failures for each policy revision, and the count of which revisions were applied on the most recent Chef Run by nodes within the segment.
* Merges the results of these queries into a single API response containing both detailed rollout information along with statistics about the Chef Client runs belonging to these rollouts.

### :chains: Related Resources

#3688 

### :athletic_shoe: How to Build and Test the Change

* Build config mgmt service and automate-gateway
* start all services (a subset probably works but I just ran them all)
* run `config_mgmt_list_rollouts_progress` to list the rollouts (this is a wrapper around `curl -fsS --insecure -H "api-token: ${api_token}" "https://localhost/api/beta/cfgmgmt/rollouts/progress_by_node_segment"` ) -- the response should be empty
* run `config_mgmt_populate_rollouts` to make sample data
* run `config_mgmt_list_rollouts_progress` to see the data.
